datafusion/functions/src/datetime/date_bin.rs

```diff
@@ -481,17 +656,82 @@ fn date_bin_impl(
+                            "DATE_BIN with Time32 source requires Time32 origin"
+                        );
+                    }
+                    let array = array.as_primitive::<Time32MillisecondType>();
+                    let apply_stride_fn = move |x: i32| {
+                        let binned_nanos =
+                            stride_fn(stride, x as i64 * NANOS_PER_MILLI, origin);
+                        let nanos = binned_nanos % (NANOSECONDS_IN_DAY);
+                        (nanos / NANOS_PER_MILLI) as i32
+                    };
+                    let array: PrimitiveArray<Time32MillisecondType> =
+                        array.unary(apply_stride_fn);
+                    ColumnarValue::Array(Arc::new(array))
+                }
+                Time32(Second) => {
+                    if !is_time {
+                        return exec_err!(
+                            "DATE_BIN with Time32 source requires Time32 origin"
+                        );
+                    }
+                    let array = array.as_primitive::<Time32SecondType>();
+                    let apply_stride_fn = move |x: i32| {
+                        let binned_nanos =
+                            stride_fn(stride, x as i64 * NANOS_PER_SEC, origin);
+                        let nanos = binned_nanos % (NANOSECONDS_IN_DAY);
+                        (nanos / NANOS_PER_SEC) as i32
+                    };
+                    let array: PrimitiveArray<Time32SecondType> =
+                        array.unary(apply_stride_fn);
+                    ColumnarValue::Array(Arc::new(array))
+                }
+                Time64(Microsecond) => {
+                    if !is_time {
+                        return exec_err!(
+                            "DATE_BIN with Time64 source requires Time64 origin"
+                        );
+                    }
+                    let array = array.as_primitive::<Time64MicrosecondType>();
+                    let apply_stride_fn = move |x: i64| {
+                        let binned_nanos = stride_fn(stride, x, origin);
+                        let nanos = binned_nanos % (NANOSECONDS_IN_DAY);
+                        nanos / NANOS_PER_MICRO
+                    };
+                    let array: PrimitiveArray<Time64MicrosecondType> =
+                        array.unary(apply_stride_fn);
+                    ColumnarValue::Array(Arc::new(array))
+                }
+                Time64(Nanosecond) => {
+                    if !is_time {
+                        return exec_err!(
+                            "DATE_BIN with Time64 source requires Time64 origin"
+                        );
+                    }
+                    let array = array.as_primitive::<Time64NanosecondType>();
+                    let apply_stride_fn = move |x: i64| {
+                        let binned_nanos = stride_fn(stride, x, origin);
+                        binned_nanos % (NANOSECONDS_IN_DAY)
```

Comment:
When the `origin` is later in the day than the `source` then the `binned_nanos` will be negative, no ?
I think you need to use https://doc.rust-lang.org/std/primitive.i64.html#method.rem_euclid instead.
Same for the similar code above.